### PR TITLE
EID-1155: Dockerize softHSM in Proxy Node Translator

### PR DESCRIPTION
**Pull request by dbes-gds**
Install softHSM in the Translator container so that the proxy node signing key can be stored in and retrieved from softHSM in test environments.
Modify startup and pki generation scripts to generate the signing key files in a format that can be loaded into softHSM.
Load the signing private key into softHSM in the Dockerfile after installing softHSM so that it is available for use during testing.
Include swedenconnect/opensaml-pkcs11 in the Translator build script so that the library can be used to access keys in softHSM.  Integration tests provided to prove that softHSM and swedenconnect/opensaml-pkcs11 are installed and usable in the container (built from translator.softHSM.Dockerfile).
The docker-compose files have been modified to use the new translator.softHSM.Dockerfile, and include additional arguments and environment variables used to load the key(s) into softHSM.
The pre-commit.sh script for Proxy Node uses a separate docker-compose file and image to compile and test the code.  This container does not use the individual proxy node component containers to run the tests, meaning that the softHSM installation was missing for pre-commit.  As such a new Dockerfile has been created to build this 'docker-gradle' service (pre-commit.softHSM.Dockerfile) which includes the softHSM installation too.  This means that the softHSM integration tests are still able to run under pre-commit.

In summary, these changes install and test softHSM and swedenconnect/opensaml-pkcs11 in the Translator Docker container, and prove the ability to load our test keys into softHSM.

The use of a key retrieved from softHSM to sign messages in the proxy node is to be implemented in a subsequent task EID-1160.

### CODE REVIEW

File: translator.softHSM.Dockerfile [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+ENTRYPOINT ["gradle", "--no-daemon"]
+
+# Runtime Image
+FROM openjdk:11-jre-slim
+
+USER root
+ENV OPEN_SC_VERSION=0.19.0
+
+# Install softHSM and OpenSC
+RUN apt-get update && apt-get install -y softhsm pcscd libccid libpcsclite-dev libssl-dev libreadline-dev autoconf automake build-essential docbook-xsl xsltproc libtool pkg-config wget && \
```
Comment (tomrosier) [inline translator.softHSM.Dockerfile]: Having all the commands compacted together will potentially make docker slower as it has todo all theses tasks every time the image is rebuilt.

File: translator.softHSM.Dockerfile [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+USER root
+
+ARG component
+COPY --from=build /app/${component}/build/install/${component} /app
+
+# Copy test signing keys
+RUN mkdir /app/keys
+COPY --from=build /app/${component}/src/dist/proxy_node_signing.* /app/keys/
+
+# softHSM Environment Variable assignment from ARGS
```
Comment (tomrosier) [inline translator.softHSM.Dockerfile]: This stuff should be at the top really, as its unlikely to change. Which will speed up builds

File: gradle/wrapper/gradle-wrapper.properties [hunk 3]
```diff
@@ -1,5 +1,6 @@
+#Tue Dec 11 10:37:58 GMT 2018
 distributionBase=GRADLE_USER_HOME
 distributionPath=wrapper/dists
-distributionUrl=https\://services.gradle.org/distributions/gradle-5.0-bin.zip
 zipStoreBase=GRADLE_USER_HOME
 zipStorePath=wrapper/dists
+distributionUrl=https\://services.gradle.org/distributions/gradle-5.0-all.zip
```
[the remark this answers is not in the document]
Comment (dbes-gds) [reply gradle/wrapper/gradle-wrapper.properties]: The 'all' version of gradle includes extra debugging tools for the ide.  Have reset this to 'bin'.

File: pre-commit.softHSM.Dockerfile [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+# Docker Image for pre-commit.sh to run all tests.
+# softHSM tests require softHSM and OpenSC.
+#
+FROM gradle:jdk11
+USER root
+
+ENV OPEN_SC_VERSION=0.19.0
+
+# Install softHSM and OpenSC
+RUN apt-get update && apt-get install -y softhsm pcscd libccid libpcsclite-dev libssl-dev libreadline-dev autoconf automake build-essential docbook-xsl xsltproc libtool pkg-config wget && \
```
[the remark this answers is not in the document]
Comment (dbes-gds) [reply pre-commit.softHSM.Dockerfile]: Have split into two separate RUNS, one for libs and one for OpenSC.  Don't think we can gain much more as this image will be fairly static and won't be rebuilt even if the code changes.

File: translator.softHSM.Dockerfile (hunk 1 above)
Comment (dbes-gds) [reply translator.softHSM.Dockerfile]: Have created a separate docker image for the softhsm installation and included it as the runtime image so that this installation does not have to happen every time there is a build.

File: translator.softHSM.Dockerfile (hunk 2 above)
Comment (dbes-gds) [reply translator.softHSM.Dockerfile]: Improve by use of the softhsm-only image.

File: pki/options.rb [hunk 5]
```diff
@@ -2,7 +2,7 @@
 
 USAGE = 'Usage: generate.rb [options] <output directory>'
 
-Options = Struct.new(:hub_entity_id, :idp_entity_id, :proxy_entity_id, :hub_response_url, :idp_sso_url, :proxy_sso_url, :do_files, :do_manifests, :do_env, :xmlsectool_path, :truststore_pass)
+Options = Struct.new(:hub_entity_id, :idp_entity_id, :proxy_entity_id, :hub_response_url, :idp_sso_url, :proxy_sso_url, :do_files, :do_manifests, :do_env, :xmlsectool_path, :truststore_pass, :softhsm)
```
[the remark this answers is not in the document]
Comment (dbes-gds) [reply pki/options.rb]: Apparently not :)

File: translator.softHSM.Dockerfile [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+COPY build.gradle build.gradle
+COPY settings.gradle settings.gradle
+COPY proxy-node-shared/ proxy-node-shared/
+
+ARG component=proxy-node-gateway
+COPY ${component}/src ${component}/src
+COPY ${component}/build.gradle ${component}/build.gradle
+
+RUN gradle --no-daemon --quiet -p ${component} --parallel install -x test
+ENTRYPOINT ["gradle", "--no-daemon"]
```
Comment (Wynndow) [inline translator.softHSM.Dockerfile]: From this line up, and from line 58 down, the Dockerfiles are the same as the base Dockerfile used to build the other images. Instead of the duplication, could you extend that Dockerfile with the extra bits here?

File: softHSM.runtime.Dockerfile [hunk 7]
```diff
@@ -0,0 +1,21 @@
+#
```
[the remark this answers is not in the document]
Comment (dbes-gds) [reply softHSM.runtime.Dockerfile]: The pre-commit one needs gradle to build whereas the runtime one doesn't, means runtime can be slimmer.

File: proxy-node-translator/src/test/java/uk/gov/ida/notification/softhsm/SoftHSMIntegrationTest.java [hunk 8]
```diff
@@ -0,0 +1,171 @@
...
+    /**
+     * softHSM tests should only run if softHSM is installed and configured.
+     * Installation and configuration to be indicated by the SOFT_HSM_LIB_PATH environment variable.
+     * softHSM is installed in the Translator Docker image and this variable is set, so the tests
+     * will run on pre-commit.  Running the tests when building locally is effectively excluded
+     * by not setting this environment variable.
+     */
+    @Before
+    public void checkSoftHSMConfigured() {
+        org.junit.Assume.assumeTrue(softHSMLibPath != null && softHSMLibPath.length() > 0);
```
[the remark this answers is not in the document]
Comment (dbes-gds) [reply proxy-node-translator/src/test/java/uk/gov/ida/notification/softhsm/SoftHSMIntegrationTest.java]: fixed

File: pre-commit.softHSM.Dockerfile [hunk 9]
```diff
@@ -0,0 +1,21 @@
+#
+# Docker Image for pre-commit.sh to run all tests.
+# softHSM tests require softHSM and OpenSC.
+# Gradle required for code build.
+#
+FROM gradle:jdk11
```
Comment (Wynndow) [inline pre-commit.softHSM.Dockerfile]: This Dockerfile is the same as the runtime one, apart from this line. Is there any way to reduce the replication?

File: build.gradle [hunk 10]
```diff
@@ -82,6 +84,10 @@ subprojects {
                 "uk.gov.ida:security-utils:${utils_version}"
         )
 
+        soft_hsm (
```
[the remark this answers is not in the document]
Comment (dbes-gds) [reply build.gradle]: tried it and it didn't work, the reference to the configuration is in the translator gradle, but the configuration definition had to go in the top level.  It kind of makes sense as that is where all the versions are defined.

File: proxy-node-translator/src/test/java/uk/gov/ida/notification/softhsm/SoftHSMIntegrationTest.java [hunk 11]
```diff
@@ -0,0 +1,171 @@
...
+
+
+    @Test
+    public void shouldRetrieveCredentialsFromHSM() throws Exception {
+
+        PKCS11ProviderConfiguration configuration = createPKCS11SoftHSMProviderConfigurationWithCredentials(CRED_ALIAS, CRED_ENV, CRED_PIN, "162624334", CRED_NAME);
+        PKCS11Provider pkcs11Provider = getPkcs11ProviderForConfiguration(configuration);
+        TestKeyPair signingTestKeyPair = new TestKeyPair(PROXY_NODE_SIGNING_CRT, PROXY_NODE_SIGNING_PK_8);
+
+        addPrivateKeyToKeyStore(signingTestKeyPair, CRED_ALIAS, CRED_PIN);
```
Comment (Wynndow) [inline proxy-node-translator/src/test/java/uk/gov/ida/notification/softhsm/SoftHSMIntegrationTest.java]: What's happening here? `addPrivateKeyToKeyStore ` doesn't seem to return anything - does it  mutate the `signingTestKeyPair` in some way?

File: pre-commit.softHSM.Dockerfile (hunk 9 above)
Comment (Wynndow) [reply pre-commit.softHSM.Dockerfile]: I think you could reduce the two files to one. You could set an `ARG` before the `FROM` that you set to the image you want to base them off (either `openjdk:11-jre-slim`, or  `gradle:jdk11`), and then use that arg in the `FROM` directive. You just need to set the arg to the right thing in the docker-compose and gradle-compose files.

File: gradle-compose.yaml [hunk 12]
```diff
@@ -2,12 +2,17 @@ version: '3'
...
+      context: .
+      dockerfile: softHSM.baseImage.Dockerfile
+      args:
+        baseImage: gradle:jdk11
     working_dir: /home/gradle/project
     user: root
     entrypoint: gradle --no-daemon
     environment:
       VERIFY_USE_PUBLIC_BINARIES: 'true'
+      SOFT_HSM_LIB_PATH: /usr/lib/softhsm/libsofthsm2.so
```
Comment (tomrosier) [inline gradle-compose.yaml]: Is this likely to change?

File: docker-compose.yaml [hunk 13]
```diff
@@ -47,9 +56,17 @@ services:
       CONNECTOR_NODE_METADATA_URL: http://stub-connector/Metadata
       CONNECTOR_NODE_ENTITY_ID: http://localhost:5000/Metadata
       REDIS_SERVER_URI: redis://gateway-redis:6379/
+      SOFT_HSM_LIB_PATH: /usr/lib/softhsm/libsofthsm2.so
```
Comment (tomrosier) [inline docker-compose.yaml]: Is this likely to change?, Can we put it in the docker file if it will stay the same 🙂
